risc0/circuit/rv32im/src/execute/executor.rs

```diff
@@ -143,17 +150,29 @@ fn unlikely(b: bool) -> bool {
     b
 }
 
-fn compute_partial_images(
-    recv: std::sync::mpsc::Receiver<ComputePartialImageRequest>,
+fn create_segments(
+    initial_image: MemoryImage,
+    recv: std::sync::mpsc::Receiver<CreateSegmentRequest>,
     mut callback: impl FnMut(Segment) -> Result<()>,
-) -> Result<()> {
+) -> Result<(Digest, Digest, MemoryImage)> {
+    let mut existing_image = initial_image;
+    let initial_digest = existing_image.image_id();
+
     while let Ok(req) = recv.recv() {
-        let partial_image = compute_partial_image(req.image, req.page_indexes);
-        callback(Segment {
+        let pre_digest = existing_image.image_id();
+        let partial_image = compute_partial_image(&mut existing_image, req.page_indexes);
+
+        for (idx, page) in req.partial_image.pages {
+            existing_image.set_page(idx, page);
```

Comment:
If you want this to be even faster, what you could do here is hash all the leaves (i.e.  Pages) in parallel using Rayon. I don't know if this makes everything faster overall, but if you make the segment creation faster, then you can start proving the first segment earlier and thus making the whole Proving pipeline faster.

Reply:
I think we could speed up the hashing maybe even more just by using the GPU to do it. At this moment like you say it probably won't speed up the overall time very much, but eventually it might be a problem.